server/lib/chat/chat-service.ts

```diff
@@ -176,25 +201,34 @@ export default class ChatService {
     let channel: ChannelInfo | undefined
     let userChannelEntry: UserChannelEntry | undefined
     let message: ChatMessage
+    let error: ChatServiceError | undefined
     do {
       attempts += 1
       try {
         await transact(async client => {
           channel = await findChannelByName(channelName, client)
           if (channel) {
             if (this.state.users.has(userId) && this.state.users.get(userId)!.has(channel.id)) {
-              throw new ChatServiceError(
+              error = new ChatServiceError(
                 ChatServiceErrorCode.AlreadyJoined,
                 'Already in this channel',
               )
+              return
             }
 
             const isBanned = await isUserBannedFromChannel(channel.id, userId, client)
             if (isBanned) {
-              throw new ChatServiceError(
+              error = new ChatServiceError(
+                ChatServiceErrorCode.UserBanned,
+                'This user has been banned from this chat channel',
+              )
+              return
+            } else if (await this.banUserFromChannelIfNeeded(channel.id, userId, client)) {
```

Comment:
You can avoid duplicating the exact same error code by just combining these conditions:
```ts
if (isBanned || await this.banUserFromChannelIfNeeded(channel.id, userId, client)) { /* ... */ }
```